File: android-core/src/main/java/com/mparticle/internal/MessageHandler.java
package com.mparticle.internal;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Looper;
import android.os.Message;

import com.mparticle.MParticle;
import com.mparticle.internal.Constants.MessageKey;
import com.mparticle.internal.Constants.MessageType;
import com.mparticle.internal.MessageManager.IncrementUserAttributeMessage;
import com.mparticle.internal.database.services.MParticleDBManager;
import com.mparticle.internal.database.tables.SessionTable;
import com.mparticle.internal.messages.BaseMPMessage;
import com.mparticle.internal.messages.MPAliasMessage;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.UUID;

/* package-private */ class MessageHandler extends BaseHandler {

    private final Context mContext;

    MParticleDBManager mMParticleDBManager;

    public static final int STORE_MESSAGE = 0;
    public static final int UPDATE_SESSION_ATTRIBUTES = 1;
    public static final int UPDATE_SESSION_END = 2;
    public static final int CREATE_SESSION_END_MESSAGE = 3;
    public static final int END_ORPHAN_SESSIONS = 4;
    public static final int STORE_BREADCRUMB = 5;
    public static final int STORE_REPORTING_MESSAGE_LIST = 9;
    public static final int REMOVE_USER_ATTRIBUTE = 10;
    public static final int SET_USER_ATTRIBUTE = 11;
    public static final int INCREMENT_USER_ATTRIBUTE = 12;
    public static final int INSTALL_REFERRER_UPDATED = 13;
    public static final int CLEAR_MESSAGES_FOR_UPLOAD = 14;
    public static final int STORE_ALIAS_MESSAGE = 15;

    private final MessageManagerCallbacks mMessageManagerCallbacks;
    String mDataplanId;
    Integer mDataplanVersion;

    /**
     * for unit testing only
     */
    MessageHandler(MessageManagerCallbacks messageManager, Context context, MParticleDBManager dbManager, String dataplanId, Integer dataplanVersion) {
        mMessageManagerCallbacks = messageManager;
        mContext = context;
        mMParticleDBManager = dbManager;
        mDataplanId = dataplanId;
        mDataplanVersion = dataplanVersion;
    }

    public MessageHandler(Looper looper, MessageManagerCallbacks messageManager, Context context, MParticleDBManager dbManager, String dataplanId, Integer dataplanVersion) {
        super(looper);
        mMessageManagerCallbacks = messageManager;
        mContext = context;
        mMParticleDBManager = dbManager;
        mDataplanId = dataplanId;
        mDataplanVersion = dataplanVersion;
    }

    boolean databaseAvailable() {
        try {
            return mMParticleDBManager.getDatabase() != null;
        } catch (Exception ex) {
            Logger.error("Database unavailable.");
            return false;
        }
    }

    @Override
    public void handleMessageImpl(Message msg) {
        try {
            if (!databaseAvailable()) {
                return;
            }
            mMessageManagerCallbacks.delayedStart();
        } catch (Exception e) {
            Logger.verbose(e.toString());
        }
        switch (msg.what) {
            case STORE_MESSAGE:
                try {
                    BaseMPMessage message = (BaseMPMessage) msg.obj;
                    message.put(MessageKey.STATE_INFO_KEY, MessageManager.getStateInfo());
                    String messageType = message.getString(MessageKey.TYPE);
                    // Handle the special case of session-start by creating the
                    // session record first.
                    if (MessageType.SESSION_START.equals(messageType)) {
                        dbInsertSession(message);
                    } else {
                        mMParticleDBManager.updateSessionEndTime(message.getSessionId(), message.getLong(MessageKey.TIMESTAMP), 0);
                        message.put(Constants.MessageKey.ID, UUID.randomUUID().toString());
                    }
                    if (MessageType.ERROR.equals(messageType)) {
                        mMParticleDBManager.appendBreadcrumbs(message);
                    }
                    try {
                        mMParticleDBManager.insertMessage(mMessageManagerCallbacks.getApiKey(), message, mDataplanId, mDataplanVersion);
                    } catch (MParticleApiClientImpl.MPNoConfigException e) {
                        Logger.error("Unable to process uploads, API key and/or API Secret are missing.");
                        return;
                    }
                    mMessageManagerCallbacks.checkForTrigger(message);

                } catch (Exception e) {
                    Logger.error(e, "Error saving message to mParticle DB.");
                }
                break;
            case INSTALL_REFERRER_UPDATED:
                try {
                    mMParticleDBManager.updateSessionInstallReferrer((String) msg.obj, mMessageManagerCallbacks.getDeviceAttributes().getAppInfo(mContext, true));
                } catch (Exception e) {
                    Logger.error(e, "Error updating session attributes in mParticle DB.");
                }
                break;
            case UPDATE_SESSION_ATTRIBUTES:
                try {
                    JSONObject sessionAttributes = (JSONObject) msg.obj;
                    String sessionId = sessionAttributes.getString(MessageKey.SESSION_ID);
                    String attributes = sessionAttributes.optString(MessageKey.ATTRIBUTES);
                    if (!MPUtility.isEmpty(attributes)) {
                        mMParticleDBManager.updateSessionAttributes(sessionId, attributes);
                    }
                } catch (Exception e) {
                    Logger.error(e, "Error updating session attributes in mParticle DB.");
                }
                break;
            case UPDATE_SESSION_END:
                try {
                    InternalSession session = (InternalSession) msg.obj;
                    mMParticleDBManager.updateSessionEndTime(session.mSessionID, session.mLastEventTime, session.getForegroundTime());
                } catch (Exception e) {
                    Logger.error(e, "Error updating session end time in mParticle DB.");
                }
                break;
            case CREATE_SESSION_END_MESSAGE:
                try {
                    Map.Entry<String, Set<Long>> entry = (Map.Entry<String, Set<Long>>) msg.obj;
                    BaseMPMessage endMessage = null;
                    String sessionId = entry.getKey();
                    try {
                        endMessage = mMParticleDBManager.getSessionForSessionEndMessage(sessionId, ((MessageManager) mMessageManagerCallbacks).getLocation(), entry.getValue());
                    } catch (JSONException jse) {
                        Logger.warning("Failed to create mParticle session end message.");
                    }
                    if (endMessage != null) {
                        try {
                            Logger.verbose("Creating session end message for session ID: " + sessionId);
                            mMParticleDBManager.insertMessage(mMessageManagerCallbacks.getApiKey(), endMessage, mDataplanId, mDataplanVersion);
                            mMParticleDBManager.updateSessionStatus(sessionId, SessionTable.SessionStatus.CLOSED);
                        } catch (MParticleApiClientImpl.MPNoConfigException e) {
                            Logger.error("Unable to process uploads, API key and/or API Secret are missing.");
                            return;
                        }

                    } else {
                        Logger.error("Error creating session end, no entry for sessionId in mParticle DB.");
                    }
                    //1 means this came from ending the session
                    if (msg.arg1 == 1) {
                        mMessageManagerCallbacks.endUploadLoop();
                    }
                } catch (Exception e) {
                    Logger.error(e, "Error creating session end message in mParticle DB.");
                } finally {

                }
                break;
            case END_ORPHAN_SESSIONS:
                try {
                    Logger.verbose("Ending orphaned sessions.");
                    // Find left-over sessions that exist during startup and end them.
                    Long mpid = (Long) msg.obj;
                    List<String> sessionIds = mMParticleDBManager.getOrphanSessionIds(mMessageManagerCallbacks.getApiKey());
                    for (String sessionId : sessionIds) {
                        Map.Entry<String, Set<Long>> entry = new HashMap.SimpleEntry<String, Set<Long>>(sessionId, Collections.singleton(mpid));
                        sendMessage(obtainMessage(MessageHandler.CREATE_SESSION_END_MESSAGE, 0, 0, entry));
                    }
                } catch (MParticleApiClientImpl.MPNoConfigException ex) {
                    Logger.error("Unable to process initialization, API key and or API Secret is missing.");
                } catch (Exception e) {
                    Logger.error(e, "Error processing initialization in mParticle DB.");
                }
                break;
            case STORE_BREADCRUMB:
                try {
                    BaseMPMessage message = (BaseMPMessage) msg.obj;
                    message.put(Constants.MessageKey.ID, UUID.randomUUID().toString());
                    try {
                        mMParticleDBManager.insertBreadcrumb(message, mMessageManagerCallbacks.getApiKey());
                    } catch (MParticleApiClientImpl.MPNoConfigException ex) {
                        Logger.error("Unable to process uploads, API key and/or API Secret are missing.");
                    }
                } catch (Exception e) {
                    Logger.error(e, "Error saving breadcrumb to mParticle DB.");
                }
                break;
            case STORE_REPORTING_MESSAGE_LIST:
                try {
                    MessageManager.ReportingMpidMessage reportingMessages = (MessageManager.ReportingMpidMessage) msg.obj;
                    mMParticleDBManager.insertReportingMessages((List<JsonReportingMessage>) reportingMessages.reportingMessages, reportingMessages.mpid);
                } catch (Exception e) {
                    Logger.verbose(e, "Error while inserting reporting messages: ", e.toString());
                }
                break;
            case REMOVE_USER_ATTRIBUTE:
                try {
                    mMParticleDBManager.removeUserAttribute((MParticleDBManager.UserAttributeRemoval) msg.obj, mMessageManagerCallbacks);
                } catch (Exception e) {
                    Logger.error(e, "Error while removing user attribute: ", e.toString());
                }
                break;
            case SET_USER_ATTRIBUTE:
                try {
                    setUserAttributes((MParticleDBManager.UserAttributeResponse) msg.obj);
                } catch (Exception e) {
                    Logger.error(e, "Error while setting user attribute: ", e.toString());
                }
                break;
            case INCREMENT_USER_ATTRIBUTE:
                try {
                    IncrementUserAttributeMessage obj = (IncrementUserAttributeMessage) msg.obj;
                    incrementUserAttribute(obj);
                } catch (Exception e) {
                    Logger.error(e, "Error while incrementing user attribute: ", e.toString());
                }
                break;
            case CLEAR_MESSAGES_FOR_UPLOAD:
                mMessageManagerCallbacks.messagesClearedForUpload();
                break;
            case STORE_ALIAS_MESSAGE:
                try {
                    MPAliasMessage aliasMessage = (MPAliasMessage) msg.obj;
                    mMParticleDBManager.insertAliasRequest(aliasMessage, mMessageManagerCallbacks.getUploadSettings());

                    MParticle instance = MParticle.getInstance();
                    if (instance != null) {
                        instance.upload();
                    }
                } catch (MParticleApiClientImpl.MPNoConfigException ex) {
                    Logger.error("Unable to Alias Request, API key and or API Secret is missing");
                } catch (Exception ex) {
                    Logger.error("Error sending Alias Request");
                }
                break;
        }
    }

    void setUserAttributes(MParticleDBManager.UserAttributeResponse response) {
        List<MParticleDBManager.AttributionChange> attributionChanges = mMParticleDBManager.setUserAttribute(response);
        for (MParticleDBManager.AttributionChange attributionChange : attributionChanges) {
            logUserAttributeChanged(attributionChange);
        }
    }

    private void incrementUserAttribute(IncrementUserAttributeMessage message) {
        Map<String, Object> userAttributes = mMParticleDBManager.getUserAttributeSingles(message.mpid);

        if (!userAttributes.containsKey(message.key)) {
            TreeMap<String, List<String>> userAttributeList = mMParticleDBManager.getUserAttributeLists(message.mpid);
            if (userAttributeList.containsKey(message.key)) {
                Logger.error("Error while attempting to increment user attribute - existing attribute is a list, which can't be incremented.");
                return;
            }
        }
        String newValue = null;
        Object currentValue = userAttributes.get(message.key);
        if (currentValue == null) {
            newValue = message.incrementBy.toString();
        } else if (currentValue instanceof Number) {
            newValue = MPUtility.addNumbers((Number) currentValue, message.incrementBy).toString();
            Logger.info("incrementing attribute: \"" + message.key + "\" from: " + currentValue + " by: " + message.incrementBy + " to: " + newValue);
        }
        MParticleDBManager.UserAttributeResponse wrapper = new MParticleDBManager.UserAttributeResponse();
        wrapper.attributeSingles = new HashMap<>(1);
        wrapper.attributeSingles.put(message.key, newValue);
        wrapper.mpId = message.mpid;
        List<MParticleDBManager.AttributionChange> attributionChanges = mMParticleDBManager.setUserAttribute(wrapper);
        for (MParticleDBManager.AttributionChange attributeChange : attributionChanges) {
            logUserAttributeChanged(attributeChange);
        }
        MParticle instance = MParticle.getInstance();
        if (instance != null && instance.Internal().getKitManager() != null) {
            instance.Internal().getKitManager().incrementUserAttribute(message.key, message.incrementBy, newValue, message.mpid);
        }
    }

    private void dbInsertSession(BaseMPMessage message) throws JSONException {
        try {
            DeviceAttributes deviceAttributes = mMessageManagerCallbacks.getDeviceAttributes();
            mMParticleDBManager.insertSession(message, mMessageManagerCallbacks.getApiKey(), deviceAttributes.getAppInfo(mContext), deviceAttributes.getDeviceInfo(mContext));
        } catch (MParticleApiClientImpl.MPNoConfigException ex) {
            Logger.error("Unable to process uploads, API key and/or API Secret are missing.");
        }
    }

    private void logUserAttributeChanged(MParticleDBManager.AttributionChange attributionChange) {
        mMessageManagerCallbacks.logUserAttributeChangeMessage(
                attributionChange.getKey(),
                attributionChange.getNewValue(),
                attributionChange.getOldValue(),
                attributionChange.isDeleted(),
                attributionChange.isNewAttribute(),
                attributionChange.getTime(),
                attributionChange.getMpId());
    }

}

File: android-core/src/main/java/com/mparticle/internal/MessageHandler.kt
package com.mparticle.internal

import android.content.Context
import android.os.Looper
import android.os.Message
import com.mparticle.MParticle
import com.mparticle.internal.Constants.MessageKey
import com.mparticle.internal.MPUtility.addNumbers
import com.mparticle.internal.MPUtility.isEmpty
import com.mparticle.internal.MParticleApiClientImpl.MPNoConfigException
import com.mparticle.internal.MessageManager.IncrementUserAttributeMessage
import com.mparticle.internal.MessageManager.ReportingMpidMessage
import com.mparticle.internal.database.services.MParticleDBManager
import com.mparticle.internal.database.services.MParticleDBManager.AttributionChange
import com.mparticle.internal.database.services.MParticleDBManager.UserAttributeRemoval
import com.mparticle.internal.database.services.MParticleDBManager.UserAttributeResponse
import com.mparticle.internal.database.tables.SessionTable
import com.mparticle.internal.messages.BaseMPMessage
import com.mparticle.internal.messages.MPAliasMessage
import org.json.JSONException
import org.json.JSONObject
import java.util.AbstractMap
import java.util.UUID

/* package-private */
internal open class MessageHandler : BaseHandler {
    private val mContext: Context

    var mMParticleDBManager: MParticleDBManager

    private val mMessageManagerCallbacks: MessageManagerCallbacks
    var mDataplanId: String? = null
    var mDataplanVersion: Int? = null

    /**
     * for unit testing only
     */
    constructor(
        messageManager: MessageManagerCallbacks,
        context: Context,
        dbManager: MParticleDBManager,
        dataplanId: String?,
        dataplanVersion: Int?
    ) {
        mMessageManagerCallbacks = messageManager
        mContext = context
        mMParticleDBManager = dbManager
        mDataplanId = dataplanId
        mDataplanVersion = dataplanVersion
    }

    constructor(
        looper: Looper,
        messageManager: MessageManagerCallbacks,
        context: Context,
        dbManager: MParticleDBManager,
        dataplanId: String?,
        dataplanVersion: Int?
    ) : super(looper) {
        mMessageManagerCallbacks = messageManager
        mContext = context
        mMParticleDBManager = dbManager
        mDataplanId = dataplanId
        mDataplanVersion = dataplanVersion
    }

    open fun databaseAvailable(): Boolean {
        try {
            return mMParticleDBManager.database != null
        } catch (ex: Exception) {
            Logger.error("Database unavailable.")
            return false
        }
    }

    override fun handleMessageImpl(msg: Message) {
        try {
            if (!databaseAvailable()) {
                return
            }
            mMessageManagerCallbacks.delayedStart()
        } catch (e: Exception) {
            Logger.verbose(e.toString())
        }
        when (msg.what) {
            STORE_MESSAGE -> try {
                val message = msg.obj as BaseMPMessage
                message.put(MessageKey.STATE_INFO_KEY, MessageManager.getStateInfo())
                val messageType = message.getString(MessageKey.TYPE)
                // Handle the special case of session-start by creating the
                // session record first.
                if (Constants.MessageType.SESSION_START == messageType) {
                    dbInsertSession(message)
                } else {
                    mMParticleDBManager.updateSessionEndTime(message.sessionId, message.getLong(MessageKey.TIMESTAMP), 0)
                    message.put(MessageKey.ID, UUID.randomUUID().toString())
                }
                if (Constants.MessageType.ERROR == messageType) {
                    mMParticleDBManager.appendBreadcrumbs(message)
                }
                try {
                    mMParticleDBManager.insertMessage(mMessageManagerCallbacks.apiKey, message, mDataplanId, mDataplanVersion)
                } catch (e: MPNoConfigException) {
                    Logger.error("Unable to process uploads, API key and/or API Secret are missing.")
                    return
                }
                mMessageManagerCallbacks.checkForTrigger(message)
            } catch (e: Exception) {
                Logger.error(e, "Error saving message to mParticle DB.")
            }

            INSTALL_REFERRER_UPDATED -> try {
                mMParticleDBManager.updateSessionInstallReferrer(
                    msg.obj as String,
                    mMessageManagerCallbacks.deviceAttributes.getAppInfo(mContext, true)
                )
            } catch (e: Exception) {
                Logger.error(e, "Error updating session attributes in mParticle DB.")
            }

            UPDATE_SESSION_ATTRIBUTES -> try {
                val sessionAttributes = msg.obj as JSONObject
                val sessionId = sessionAttributes.getString(MessageKey.SESSION_ID)
                val attributes = sessionAttributes.optString(MessageKey.ATTRIBUTES)
                if (!isEmpty(attributes)) {
                    mMParticleDBManager.updateSessionAttributes(sessionId, attributes)
                }
            } catch (e: Exception) {
                Logger.error(e, "Error updating session attributes in mParticle DB.")
            }

            UPDATE_SESSION_END -> try {
                val session = msg.obj as InternalSession
                mMParticleDBManager.updateSessionEndTime(session.mSessionID, session.mLastEventTime, session.foregroundTime)
            } catch (e: Exception) {
                Logger.error(e, "Error updating session end time in mParticle DB.")
            }

            CREATE_SESSION_END_MESSAGE -> try {
                val entry = msg.obj as Map.Entry<String, Set<Long>>
                var endMessage: BaseMPMessage? = null
                val sessionId = entry.key
                try {
                    endMessage = mMParticleDBManager.getSessionForSessionEndMessage(
                        sessionId,
                        (mMessageManagerCallbacks as MessageManager).location,
                        entry.value
                    )
                } catch (jse: JSONException) {
                    Logger.warning("Failed to create mParticle session end message.")
                }
                if (endMessage != null) {
                    try {
                        Logger.verbose("Creating session end message for session ID: $sessionId")
                        mMParticleDBManager.insertMessage(mMessageManagerCallbacks.apiKey, endMessage, mDataplanId, mDataplanVersion)
                        mMParticleDBManager.updateSessionStatus(sessionId, SessionTable.SessionStatus.CLOSED)
                    } catch (e: MPNoConfigException) {
                        Logger.error("Unable to process uploads, API key and/or API Secret are missing.")
                        return
                    }
                } else {
                    Logger.error("Error creating session end, no entry for sessionId in mParticle DB.")
                }
                // 1 means this came from ending the session
                if (msg.arg1 == 1) {
                    mMessageManagerCallbacks.endUploadLoop()
                }
            } catch (e: Exception) {
                Logger.error(e, "Error creating session end message in mParticle DB.")
            } finally {
            }

            END_ORPHAN_SESSIONS -> try {
                Logger.verbose("Ending orphaned sessions.")
                // Find left-over sessions that exist during startup and end them.
                val mpid = msg.obj as Long
                val sessionIds = mMParticleDBManager.getOrphanSessionIds(mMessageManagerCallbacks.apiKey)
                for (sessionId in sessionIds) {
                    val entry: Map.Entry<String, Set<Long>> = AbstractMap.SimpleEntry(sessionId, setOf(mpid))
                    sendMessage(obtainMessage(CREATE_SESSION_END_MESSAGE, 0, 0, entry))
                }
            } catch (ex: MPNoConfigException) {
                Logger.error("Unable to process initialization, API key and or API Secret is missing.")
            } catch (e: Exception) {
                Logger.error(e, "Error processing initialization in mParticle DB.")
            }

            STORE_BREADCRUMB -> try {
                val message = msg.obj as BaseMPMessage
                message.put(MessageKey.ID, UUID.randomUUID().toString())
                try {
                    mMParticleDBManager.insertBreadcrumb(message, mMessageManagerCallbacks.apiKey)
                } catch (ex: MPNoConfigException) {
                    Logger.error("Unable to process uploads, API key and/or API Secret are missing.")
                }
            } catch (e: Exception) {
                Logger.error(e, "Error saving breadcrumb to mParticle DB.")
            }

            STORE_REPORTING_MESSAGE_LIST -> try {
                val reportingMessages = msg.obj as ReportingMpidMessage
                mMParticleDBManager.insertReportingMessages(
                    reportingMessages.reportingMessages as List<JsonReportingMessage?>,
                    reportingMessages.mpid
                )
            } catch (e: Exception) {
                Logger.verbose(e, "Error while inserting reporting messages: ", e.toString())
            }

            REMOVE_USER_ATTRIBUTE -> try {
                mMParticleDBManager.removeUserAttribute(msg.obj as UserAttributeRemoval, mMessageManagerCallbacks)
            } catch (e: Exception) {
                Logger.error(e, "Error while removing user attribute: ", e.toString())
            }

            SET_USER_ATTRIBUTE -> try {
                setUserAttributes(msg.obj as UserAttributeResponse)
            } catch (e: Exception) {
                Logger.error(e, "Error while setting user attribute: ", e.toString())
            }

            INCREMENT_USER_ATTRIBUTE -> try {
                val obj = msg.obj as IncrementUserAttributeMessage
                incrementUserAttribute(obj)
            } catch (e: Exception) {
                Logger.error(e, "Error while incrementing user attribute: ", e.toString())
            }

            CLEAR_MESSAGES_FOR_UPLOAD -> mMessageManagerCallbacks.messagesClearedForUpload()
            STORE_ALIAS_MESSAGE -> try {
                val aliasMessage = msg.obj as MPAliasMessage
                mMParticleDBManager.insertAliasRequest(mMessageManagerCallbacks.apiKey, aliasMessage)

                val instance = MParticle.getInstance()
                instance?.upload()
            } catch (ex: MPNoConfigException) {
                Logger.error("Unable to Alias Request, API key and or API Secret is missing")
            } catch (ex: Exception) {
                Logger.error("Error sending Alias Request")
            }
        }
    }

    fun setUserAttributes(response: UserAttributeResponse?) {
        val attributionChanges = mMParticleDBManager.setUserAttribute(response)
        for (attributionChange in attributionChanges) {
            logUserAttributeChanged(attributionChange)
        }
    }

    private fun incrementUserAttribute(message: IncrementUserAttributeMessage) {
        val userAttributes = mMParticleDBManager.getUserAttributeSingles(message.mpid)

        if (!userAttributes.containsKey(message.key)) {
            val userAttributeList = mMParticleDBManager.getUserAttributeLists(message.mpid)
            if (userAttributeList.containsKey(message.key)) {
                Logger.error("Error while attempting to increment user attribute - existing attribute is a list, which can't be incremented.")
                return
            }
        }
        var newValue: String? = null
        val currentValue = userAttributes[message.key]
        if (currentValue == null) {
            newValue = message.incrementBy.toString()
        } else if (currentValue is Number) {
            newValue = addNumbers(currentValue, message.incrementBy).toString()
            Logger.info("incrementing attribute: \"" + message.key + "\" from: " + currentValue + " by: " + message.incrementBy + " to: " + newValue)
        }
        val wrapper = UserAttributeResponse()
        wrapper.attributeSingles = HashMap(1)
        wrapper.attributeSingles[message.key] = newValue
        wrapper.mpId = message.mpid
        val attributionChanges = mMParticleDBManager.setUserAttribute(wrapper)
        for (attributeChange in attributionChanges) {
            logUserAttributeChanged(attributeChange)
        }
        val instance = MParticle.getInstance()
        if (instance != null && instance.Internal().kitManager != null) {
            instance.Internal().kitManager.incrementUserAttribute(message.key, message.incrementBy, newValue, message.mpid)
        }
    }

    @Throws(JSONException::class)
    private fun dbInsertSession(message: BaseMPMessage) {
        try {
            val deviceAttributes = mMessageManagerCallbacks.deviceAttributes
            mMParticleDBManager.insertSession(
                message,
                mMessageManagerCallbacks.apiKey,
                deviceAttributes.getAppInfo(mContext),
                deviceAttributes.getDeviceInfo(mContext)
            )
        } catch (ex: MPNoConfigException) {
            Logger.error("Unable to process uploads, API key and/or API Secret are missing.")
        }
    }

    private fun logUserAttributeChanged(attributionChange: AttributionChange) {
        mMessageManagerCallbacks.logUserAttributeChangeMessage(
            attributionChange.key,
            attributionChange.newValue,
            attributionChange.oldValue,
            attributionChange.isDeleted,
            attributionChange.isNewAttribute,
            attributionChange.time,
            attributionChange.mpId
        )
    }

    companion object {
        const val STORE_MESSAGE: Int = 0
        const val UPDATE_SESSION_ATTRIBUTES: Int = 1
        const val UPDATE_SESSION_END: Int = 2
        const val CREATE_SESSION_END_MESSAGE: Int = 3
        const val END_ORPHAN_SESSIONS: Int = 4
        const val STORE_BREADCRUMB: Int = 5
        const val STORE_REPORTING_MESSAGE_LIST: Int = 9
        const val REMOVE_USER_ATTRIBUTE: Int = 10
        const val SET_USER_ATTRIBUTE: Int = 11
        const val INCREMENT_USER_ATTRIBUTE: Int = 12
        const val INSTALL_REFERRER_UPDATED: Int = 13
        const val CLEAR_MESSAGES_FOR_UPLOAD: Int = 14
        const val STORE_ALIAS_MESSAGE: Int = 15
    }
}
